src/lib/cpd/CpdEngine.ts

```diff
@@ -0,0 +1,246 @@
+import * as engineUtils from '../util/CommonEngineUtils';
+import CpdWrapper from './CpdWrapper';
+import {FILE_EXTS_TO_LANGUAGE} from '../../Constants';
+import { uxEvents } from '../ScannerEvents';
+
+Messages.importMessagesDirectory(__dirname);
+const messages = Messages.loadMessages('@salesforce/sfdx-scanner', 'CpdEngine');
+
+export class CpdEngine extends AbstractRuleEngine {
+
+	public static ENGINE_ENUM: ENGINE = ENGINE.CPD;
+	public static ENGINE_NAME: string = ENGINE.CPD.valueOf();
+	private static minimumTokens: number;
+
+	private logger: Logger;
+	private config: Config;
+	private initialized: boolean;
+	private cpdCatalog: Catalog;
+	
+
+
+	public getName(): string {
+		return CpdEngine.ENGINE_NAME;
+	}
+
+	public getTargetPatterns(): Promise<TargetPattern[]> {
+		return this.config.getTargetPatterns(ENGINE.CPD);
+	}
+
+	public getCatalog(): Promise<Catalog>{
+		return Promise.resolve(this.cpdCatalog);
+	}
+
+	/* eslint-disable-next-line no-unused-vars, @typescript-eslint/no-unused-vars */
+	public shouldEngineRun(ruleGroups: RuleGroup[], rules: Rule[], target: RuleTarget[], engineOptions: Map<string, string>): boolean {
+		return true;
+	}
+	
+	// eslint-disable-next-line @typescript-eslint/no-unused-vars
+	public async run(ruleGroups: RuleGroup[], rules: Rule[], targets: RuleTarget[], engineOptions: Map<string, string>): Promise<RuleResult[]> {
+
+		const languageToPaths = this.sortPaths(targets);
+		const results = [];
+
+		if (languageToPaths.size === 0) {
+			this.logger.trace(`No matching cpd target files found. Nothing to execute.`);
+			return [];
+		}
+		
+		// cpd only allows one language to be speificed at a time, so we must run it for each
+		for (const language of Array.from(languageToPaths.keys())) {
+			try {
+				const stdout = await this.runLanguage(language, languageToPaths.get(language));
+				const languageResults = this.processStdOut(stdout)
+				results.push(...languageResults);
+				this.logger.trace(`Found ${languageResults.length} for CPD (${language})`);
+
+			} catch (e) {
+				this.logger.trace(`Cpd (${language}) evaluation failed: ` + (e.message || e));
+				throw new SfdxError(e.message || e);
+			}
+		}   
+		return results;
+	}
+
+	private sortPaths(targets: RuleTarget[]): Map<string, string[]> {
+		const languageToPaths = new Map();
+
+		for (const target of targets) {
+			for (const path of target.paths) {
+				const i = path.lastIndexOf(".");
+				if (i === -1) {
+					//this.logger.warn(`Target: '${path}' was not processed by CPD, no file extension found.`)
+					
+					uxEvents.emit('warning-always', `Target: '${path}' was not processed by CPD, no file extension found.`);
+					continue;
+				}
+				const ext = path.substr(i);
+				if (FILE_EXTS_TO_LANGUAGE.has(ext)) {
+					const language = FILE_EXTS_TO_LANGUAGE.get(ext);
+					const currentPaths = [];
+					if (languageToPaths.has(language)) {
+						currentPaths.push(...languageToPaths.get(language));	
```

Comment:
Feels a little inefficient to create a new array every time only to push one more item to an existing array. How about getting the existing array from `languageToPaths` and doing a `push` with the new item to add? The map only holds the reference to the array and shouldn't mind when the array gets modified.

Reply:
Didn't realize I could directly push onto the map value, thanks!